target/riscv/cpu_helper.c

```diff
@@ -716,11 +716,13 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
             if ((pte & PTE_X)) {
                 *prot |= PAGE_EXEC;
             }
-            /* add write permission on stores or if the page is already dirty,
-               so that we TLB miss on later writes to update the dirty bit */
-            if ((pte & PTE_W) &&
-                ((access_type == MMU_DATA_STORE) ||
-                 (access_type == MMU_DATA_CAP_STORE) || (pte & PTE_D))) {
+            /*
+             * Require that the page is both writable and dirty in order to
+             * populate the TLB with a store-permitting entry.  If either of
+             * PTE_W or PTE_D are clear, force a TLB miss on later writes to
+             * update the dirty bit
+             */
+            if ((pte & PTE_W) && (pte & PTE_D)) {
```

Comment:
This diverges from upstream, where it's:
```
            if ((pte & PTE_W) &&
                    (access_type == MMU_DATA_STORE || (pte & PTE_D))) {
```
though I don't think it's ever possible to get here with `access_type == MMU_DATA_(CAP_)STORE` but not `pte & PTE_D`, as either you hit the "atomically set PTE_D" case or you hit one of your RISCV_PTE_TRAPPY cases and bail out early. I therefore think this is an entirely no-op change and so we should stick to extending upstream's approach.

Reply:
Yes, this is a no-op change because `access_type == MMU_DATA_STORE` implies `pte & PTE_D`.  I can revert to upstream if that'd make you happier.